### PR TITLE
support json format for all_subnets analysis type

### PR DESCRIPTION
**Pull request by adisos**

### CODE REVIEW

**Review by kyorav — APPROVED** [review]
output looks good

**Comment by kyorav** [comment]
I would like to suggest that we align the format for all_subnets with that of all_endpoints (see also change request for the all_endpoints analysis in issue #143 ).
This means that `src` and `dst` would be objects, with the same fields as for the all_endpoints analysis and adding a new type "subnet".
This would greatly help implementing checks over the json output.

File: pkg/common/connectionset.go [hunk 1]
```diff
@@ -343,6 +343,121 @@ func (conn *ConnectionSet) String() string {
 	return strings.Join(resStrings, "; ")
 }
 
+type ConnDetails ProtocolList /*struct {
+	AllowedConns ProtocolList
+}*/
```
Comment (zivnevo) [inline pkg/common/connectionset.go:348]: Remove commented-out code

File: pkg/common/connectionset.go [hunk 2]
```diff
@@ -343,6 +343,121 @@ func (conn *ConnectionSet) String() string {
...
+	return res
+}
+
+func ConnToJSONRep(c *ConnectionSet) ConnDetails {
+	res := ProtocolList{}
+	if c.AllowAll {
+		res = append(res, AnyProtocol{Protocol: AnyProtocolProtocolANY})
+		return ConnDetails(res)
+	}
+
```
Comment (zivnevo) [inline pkg/common/connectionset.go:434]:
```suggestion
	if c.AllowAll {
		return ConnDetails(ProtocolList{AnyProtocol{Protocol: AnyProtocolProtocolANY}})
	}

	res := ProtocolList{}
```

**Review by zivnevo — APPROVED** [review]
I have no comment that should block merging this PR.
Please see my questions and suggestions below - maybe for future work...

File: pkg/common/data_model.go
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/common/data_model.go:1]: yes

File: pkg/ibmvpc/analysis_output_test.go [hunk 3]
```diff
@@ -238,15 +238,24 @@ var tests = []*vpcGeneralTest{
...
 	},
+	// json example
+	{
+		name:     "demo_with_instances",
+		useCases: []vpcmodel.OutputUseCase{vpcmodel.AllEndpoints, vpcmodel.AllSubnets},
+		format:   vpcmodel.JSON,
+	},
 }
 
+var formatsAvoidComparison = map[vpcmodel.OutFormat]bool{vpcmodel.ARCHDRAWIO: true, vpcmodel.DRAWIO: true, vpcmodel.JSON: true}
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/analysis_output_test.go:249]: The json output is currently not sorted

**Comment by adisos** [comment]
> I have no comment that should block merging this PR. Please see my questions and suggestions below - maybe for future work...

Opened a new PR to add updates based on the comments: https://github.com/np-guard/vpc-network-config-analyzer/pull/159

File: pkg/ibmvpc/parser.go [hunk 4]
```diff
@@ -310,17 +310,19 @@ func getInstancesConfig(
 	for i := range instanceList {
 		instance := instanceList[i]
 		vsiNode := &Vsi{
-			VPCResource: vpcmodel.VPCResource{ResourceName: *instance.Name, ResourceUID: *instance.CRN, Zone: *instance.Zone.Name},
-			nodes:       []vpcmodel.Node{},
+			VPCResource: vpcmodel.VPCResource{ResourceName: *instance.Name, ResourceUID: *instance.CRN, Zone: *instance.Zone.Name,
+				ResourceType: "VSI"},
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go:314]: done

File: pkg/vpcmodel/externalNetwork.go [hunk 5]
```diff
@@ -96,20 +95,19 @@ func getPublicInternetIPblocksList() (internetIPblocksList []*common.IPBlock, al
...
 	if isPublicInternet {
 		return &ExternalNetwork{
-			VPCResource: VPCResource{ResourceName: publicInternetNodeName},
+			VPCResource: VPCResource{ResourceType: publicInternetNodeName},
 			CidrStr:     cidr, isPublicInternet: true,
 		}, nil
 	}
-	nodeName := fmt.Sprintf("ref-address-%d", index)
-	return &ExternalNetwork{VPCResource: VPCResource{ResourceName: nodeName}, CidrStr: cidr}, nil
+	return &ExternalNetwork{VPCResource: VPCResource{ResourceType: publicInternetNodeName}, CidrStr: cidr}, nil
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/externalNetwork.go:110]: done

File: pkg/vpcmodel/jsonOutput.go [hunk 6]
```diff
@@ -100,7 +93,35 @@ func (j *JSONoutputFormatter) WriteOutputAllEndpoints(c *CloudConfig, conn *VPCC
...
+}
+
+func getConnLinesForSubnetsConnectivity(conn *VPCsubnetConnectivity) []subnetsConnectivityConnLine {
+	connLines := []subnetsConnectivityConnLine{}
+	for src, nodeConns := range conn.AllowedConnsCombined {
+		for dst, conns := range nodeConns {
+			if conns.IsEmpty() {
+				continue
+			}
+			// currently not supported with grouping
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/jsonOutput.go:118]: added a check when parsing args, issue and error if json output and grouping flag are both present.

File: pkg/common/connectionset.go (hunk 1 above)
Comment (adisos) [reply pkg/common/connectionset.go:348]: done

File: pkg/common/connectionset.go (hunk 2 above)
Comment (adisos) [reply pkg/common/connectionset.go:434]: done

**Pull request merged**